### DISCUSSION TITLE
1500 Archers on a 28.8: Network Programming in Age of Empires and Beyond

### DISCUSSION THREAD

**Discussion by JonasBorchelt** [SFTtech/openage#1493] [Ideas]
Maybe this is quite interesting?

https://www.gamedeveloper.com/programming/1500-archers-on-a-28-8-network-programming-in-age-of-empires-and-beyond

**Comment by TheJJ** [comment]
This is indeed an interesting read, but this architecture relies on distributing input events between decentral clients, where all clients have all the information and just hide them from the user. The openage architecture works with a central truth (the game server) which receives input events from the clients, and then distributes the interesting part of the server gamestate to each client (possibly with client-local prediction). We decided against the decentralized architecture to have one point of truth and hence a simpler implementation, and additional features like only one point where mods apply their changes, cheating prevention and server-side prediction of the game state through curves.

**Reply by JonasBorchelt** [reply]
I can remember we already talked about how it will implemented in openage. 👍

**Comment by Bigbossbro08** [comment]
I worked a bit on this. Honestly its not worth it to go against distributing input approach. Best you can do split the deterministic simulation as a whole. Like splitting wherever its possible and relevant. Link chunking the maps, making researching tech on itself being some form of simulation. Birds or idle units, buildings being on separate simulation wherever its possible. Thus reducing workload. That is if you plan to do MMO like or sort. In this approach you can atleast transfer units with all of its data between small deterministic game worlds. It is if you care network bandwidths. Syncing individual can prove more expensive. Pretty much every other game still uses lockstep sim for a reason.
